### PR TITLE
feat: Pass sync context into all sync flows

### PR DESCRIPTION
**Pull request by qingchm**
#### Which issue(s) does this change fix?
N.A.

#### Why is this change necessary?
In order to pass sync context information into all sync flows, which enables more information to be obtained when each sync flow executes

#### How does it address the issue?
Pass in sync context object into all sync flow construction.

#### What side effects does this change have?
None.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/commands/sync/command.py [hunk 1]
```diff
@@ -348,14 +348,28 @@ def do_cli(
...
                         if watch:
                             execute_watch(
-                                template_file, build_context, package_context, deploy_context, dependency_layer, code
+                                template_file,
+                                build_context,
+                                package_context,
+                                deploy_context,
+                                dependency_layer,
+                                code,
+                                sync_context,
```
Comment (mndeveci) [inline samcli/commands/sync/command.py:362]: nit: can we combine context objects together (for other places as well) for readability?

**Review by mndeveci — COMMENTED** [review]
Thanks for the change, LGTM! Left a small comment for readability

File: samcli/commands/sync/command.py (hunk 1 above)
Comment (qingchm) [reply samcli/commands/sync/command.py]: Updated

**Review by mndeveci — APPROVED** [review]
Thanks for addressing the feedback, can you check why CI jobs are failing?

**Review by hnnasit — APPROVED** [review]
LGTM too

**Pull request merged**